include/tvm/node/container.h

```diff
@@ -43,51 +44,1188 @@ using runtime::Downcast;
+
+inline const MapNode::mapped_type& MapNode::at(const MapNode::key_type& key) const {
+  _TVM_DISPATCH_MAP_CONST(this, p, { return p->at(key); });
+}
+
+inline MapNode::mapped_type& MapNode::at(const MapNode::key_type& key) {
+  _TVM_DISPATCH_MAP(this, p, { return p->at(key); });
+}
+
+inline MapNode::iterator MapNode::begin() const {
+  _TVM_DISPATCH_MAP_CONST(this, p, { return p->begin(); });
+}
+
+inline MapNode::iterator MapNode::end() const {
+  _TVM_DISPATCH_MAP_CONST(this, p, { return p->end(); });
+}
+
+inline MapNode::iterator MapNode::find(const MapNode::key_type& key) const {
+  _TVM_DISPATCH_MAP_CONST(this, p, { return p->find(key); });
+}
+
+inline void MapNode::erase(const MapNode::iterator& position) {
+  _TVM_DISPATCH_MAP(this, p, { return p->erase(position); });
+}
+
+#undef _TVM_DISPATCH_MAP
+#undef _TVM_DISPATCH_MAP_CONST
+
+inline ObjectPtr<MapNode> MapNode::Empty() { return SmallMapNode::Empty(); }
+
+inline ObjectPtr<MapNode> MapNode::CopyFrom(MapNode* m) {
+  if (m->slots_ <= SmallMapNode::kMaxSize) {
+    return SmallMapNode::CopyFrom(static_cast<SmallMapNode*>(m));
+  } else {
+    return DenseMapNode::CopyFrom(static_cast<DenseMapNode*>(m));
+  }
+}
+
+template <typename IterType>
+inline ObjectPtr<Object> MapNode::CreateFromRange(IterType first, IterType last) {
+  int64_t _cap = std::distance(first, last);
+  if (_cap < 0) {
+    return SmallMapNode::Empty();
+  }
+  uint64_t cap = static_cast<uint64_t>(_cap);
+  if (cap < SmallMapNode::kMaxSize) {
+    return SmallMapNode::CreateFromRange(cap, first, last);
+  }
+  uint32_t fib_shift;
+  uint64_t n_slots;
+  DenseMapNode::CalcTableSize(cap, &fib_shift, &n_slots);
+  ObjectPtr<Object> n = DenseMapNode::Empty(fib_shift, n_slots);
+  for (; first != last; ++first) {
+    KVType kv(*first);
+    DenseMapNode::InsertMaybeReHash(kv, &n);
+  }
+  return n;
+}
+
+inline void MapNode::InsertMaybeReHash(const KVType& kv, ObjectPtr<Object>* map) {
+  constexpr uint64_t kMaxSize = SmallMapNode::kMaxSize;
+  MapNode* base = static_cast<MapNode*>(map->get());
+  if (base->slots_ < kMaxSize) {
+    SmallMapNode::InsertMaybeReHash(kv, map);
+    CHECK_LE(static_cast<MapNode*>(map->get())->slots_, kMaxSize);
+  } else if (base->slots_ == kMaxSize) {
+    if (base->size_ < base->slots_) {
+      SmallMapNode::InsertMaybeReHash(kv, map);
+      CHECK_LE(static_cast<MapNode*>(map->get())->slots_, kMaxSize);
+    } else {
+      ObjectPtr<Object> new_map = MapNode::CreateFromRange(base->begin(), base->end());
+      CHECK_GT(static_cast<MapNode*>(new_map.get())->slots_, kMaxSize);
+      DenseMapNode::InsertMaybeReHash(kv, &new_map);
+      *map = std::move(new_map);
+      CHECK_GT(static_cast<MapNode*>(map->get())->slots_, kMaxSize);
+    }
+  } else {
+    DenseMapNode::InsertMaybeReHash(kv, map);
+    CHECK_GT(static_cast<MapNode*>(map->get())->slots_, kMaxSize);
+  }
+}
+
+namespace runtime {
+template <>
+inline ObjectPtr<MapNode> make_object<>() {
+  return SmallMapNode::Empty();
```

Comment:
Should we simply use static assert to disable make_object for MapNode?

Reply:
I think SmallMapNode/DenseMapNode are details that we might want to hide, and hopefully ordinary users only touch MapNode. This is to help users not to mention SmallMapNode even once.

Reply:
it is still strange to call make_object<MapNode> here, perhaps we can expose MapNode::Empty instead

Reply:
Okay, here is what I am going to do:
* Mark make_object as deleted.
* Expose MapNode::Empty instead

Does it look good?

Reply:
Done :-)